Fix #5015: Added CI to oppia-android wiki (check toc)

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fixes #5015

This introduces a script to verify that the table of contents in each wiki page corresponds to its headers whenever a wiki change is made, ensuring consistency and correctness in documentation. 

Note: Ensure any emojis included in the headers are also reflected in the table of contents.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Rd4dev committed:
Script to Check the Table of Contents in the Wiki and CI workflow

Rd4dev committed:
Fix Lint Tests by adding new lines after and before ( and )

Rd4dev committed:
Fix Lint Tests by properly indenting the error messages

Rd4dev commented:
@adhiamboperes PTAL

adhiamboperes commented on .github/workflows/wiki.yml:
```diff
@@ -10,6 +10,24 @@ on:
 
 jobs:
+  table_of_contents_check:
+    name: Wiki Check Table of Contents match Headers
```
```suggestion
    name: Check Wiki Table of Contents
```

You can add more clarifying details as a comment above it.

Rd4dev replied:
Done! I've made the name concise and moved the extra info as a comment above it.

adhiamboperes commented on .github/workflows/wiki.yml:
```diff
@@ -10,6 +10,24 @@ on:
+        with:
+          version: 4.0.0
+
+      - name: Wiki Check Table of Contents
+        id: wikichecktoc
```
```suggestion
      - name: Check Wiki Table of Contents
        id: checkWikiToc
```

From looking at other IDs in other yml files, the common pattern is to use camel case.

Rd4dev replied:
Thanks for the insight, @adhiamboperes. I've replaced the ID with camel case in [wiki.yml](https://github.com/oppia/oppia-android/pull/5382/commits/d113747e709ac4755adaedaba40f8d574a8ba356). Upon further exploration, I noticed that [issue_checks.yml](https://github.com/oppia/oppia-android/blob/a0deeea74289c94797dd9d3729ee7c157030ab67/.github/workflows/issue_checks.yml#L25) follows camel case, while [static_check.yml](https://github.com/oppia/oppia-android/blob/a0deeea74289c94797dd9d3729ee7c157030ab67/.github/workflows/static_checks.yml#L110) and most IDs in [unit_tests.yml](https://github.com/oppia/oppia-android/blob/a0deeea74289c94797dd9d3729ee7c157030ab67/.github/workflows/unit_tests.yml#L161) follow snake case. Additionally, one ID in [unit_tests.yml](https://github.com/oppia/oppia-android/blob/a0deeea74289c94797dd9d3729ee7c157030ab67/.github/workflows/unit_tests.yml#L76) follows kebab case.  Should there be any changes made to follow a common convention?

adhiamboperes replied:
It would be better to file a new issue so that we can discuss an implement some standardization.

Rd4dev replied:
Got it! Will stick with camel case with this [PR](https://github.com/oppia/oppia-android/pull/5382/commits/d113747e709ac4755adaedaba40f8d574a8ba356) then and work on the tests.

adhiamboperes reviewed (changes requested):
Thanks @Rd4dev!

This mostly looks good, but you are missing tests for these validations in the  `RegexPatternValidationCheckTest.kt` file.

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

Rd4dev committed:
Modified id with camel case convention and Revised the name to be more concise

seanlip commented:
@Rd4dev Could you please add tests for this PR (and also confirm that you have managed to get Bazel tests working locally)?

/cc @BenHenning

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

Rd4dev committed:
Resolve merge conflicts and fixed kdoc styling

Rd4dev committed:
Renaming wiki dir variables for better clarity

Rd4dev committed:
Simplified the wiki table of content check logic with kotlin functions

Rd4dev committed:
Code clean up and replaced exception with error statement

Rd4dev committed:
Added test cases for wiki toc checks

Rd4dev committed:
Fix Lint Checks for indentation and missing spaces

Rd4dev committed:
Fix Lint check buildifier reformat

Rd4dev commented:
Hey @adhiamboperes, I have reopened this PR and added the test cases in `WikiTableOfContentsCheckTest.kt`. Can you PTAL?
Also, pointing out to a few existing wiki toc mismatches mentioned earlier - https://github.com/oppia/oppia-android/pull/5382#issue-2236257658

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

BenHenning commented:
What needs to happen next to finish up this PR @Rd4dev / @adhiamboperes?

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

Rd4dev commented:
> What needs to happen next to finish up this PR @Rd4dev / @adhiamboperes?

@BenHenning, not sure, just waiting for the reviews—I'll proceed once they are in!

Rd4dev committed:
Added missing test cases for no directory found to hit 100% coverage

adhiamboperes commented on .github/workflows/wiki.yml:
```diff
@@ -5,11 +5,30 @@ on:
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Bazel
+        uses: abhinavsingh/setup-bazel@v3
+        with:
+          version: 4.0.0
+
```
Please update to current version of Bazel

Rd4dev replied:
Done, Updated it!

adhiamboperes commented on .github/workflows/wiki.yml:
```diff
@@ -5,11 +5,30 @@ on:
+    # To verify that the wiki's table of contents matches the headers accurately.
+    name: Check Wiki Table of Contents
+    runs-on: ubuntu-20.04
+    env:
+      CACHE_DIRECTORY: ~/.bazel_cache
```
Is the cache being used at all?

Rd4dev replied:
As discussed, we don't seem to utilize them at the moment, so I've removed them.

adhiamboperes commented on scripts/src/java/org/oppia/android/scripts/wiki/WikiTableOfContentsCheck.kt:
```diff
@@ -0,0 +1,100 @@
+fun checkTableOfContents(file: File) {
+  val fileContents = file.readLines()
+  val tocStartIdx = fileContents.indexOfFirst {
+    it.contains("## Table of Contents")
```
What happens if there are multiple spaces after ##? do we want to handle that?

Rd4dev replied:
Ah, I hadn't thought of that! I've now updated it to use a Regex with Ignore Case to also accept variations like 
- TABLE OF CONTENTS 
- Table of contents 
- Table Of   Contents
- etc.

adhiamboperes commented on scripts/src/java/org/oppia/android/scripts/wiki/WikiTableOfContentsCheck.kt:
```diff
@@ -0,0 +1,100 @@
+
+  val linkRegex = "\\(#(.*?)\\)".toRegex()
+  val link = linkRegex.find(line)?.groupValues?.get(1)?.removePrefix("#")?.replace('-', ' ')
+    ?.replace(Regex("[?&./:’'*!,(){}\\[\\]+]"), "")
+    ?.replace("confetti_ball", "")?.trim()
```
Might we, instead, want to ban emojis? It would reduce the maintainance burden. WDYT?

Rd4dev replied:
I have two thoughts on this but I'm unsure of the right approach:

1. We could restrict using emojis (But would this require adding a check to ensure no emojis are included in headers, or would it just be a manual review check?).
2. Alternatively, we could still accept emojis, but they should also be included in the TOC titles in the `:emoji:` form for comparison and validation.

adhiamboperes replied:
Wouldn't emojis fail by default since their presence would cause a mismatch?

Rd4dev replied:
Ah, got it. I thought we should implement something to restrict adding emojis separately while pushing. In that case, the current check could work in both ways - by checking the TOC and also by allowing contributors to add emojis (if needed) but in a certain pattern.

- If an emoji is added as a visual emoji (standard image emoji) [🦦], the headers should ignore adding it. - [This is an option if contributors want to include emojis in their headers but don't want them in the TOC representation]
- If an emoji is added as GitHub emoji code `(:emoji:)` [:otter:], the headers should include it. - [This is an option if contributors want to represent them in both headers and the TOC representation]

Few examples that would work are:

### Table of Contents
- [Instructions1](#instructions1-)
- [Instructions2 :otter:](#instructions2-otter)
- [Instructions3](#-instructions3)
- [:otter: Instructions4](#otter-instructions4)

```
- [Instructions1](#instructions1-)
- [Instructions2 :otter:](#instructions2-otter)
- [Instructions3](#-instructions3)
- [:otter: Instructions4](#otter-instructions4)
```

### Headers
- Instructions1 🦦
- Instructions2 :otter:
- 🦦 Instructions3
- :otter: Instructions4

```
## Instructions1 🦦
## Instructions2 :otter:
## 🦦 Instructions3
## :otter: Instructions4
```

So, anything in the wrong format would be checked by the script, and contributors can decide how they want to represent it, I guess.

adhiamboperes replied:
Yes, exactly. Also resolving since we agreed in the meeting discussion.

adhiamboperes commented on scripts/src/java/org/oppia/android/scripts/wiki/WikiTableOfContentsCheck.kt:
```diff
@@ -0,0 +1,100 @@
+  val eOfIdx = fileContents.subList(tocStartIdx + 2, fileContents.size).indexOfFirst {
+    it.isBlank()
+  }
+  if (eOfIdx == -1) error("Table of Contents didn't end with a blank line")
```
```suggestion
  if (eOfIdx == -1) error("Table of Contents didn't end with a blank line.")
```

Rd4dev replied:
Updated it.

adhiamboperes reviewed (changes requested):
Thanks @Rd4dev! PTAL at my comments inline.

Rd4dev committed:
Upgraded the Bazel version, removed cache env, used Regex to check heading text

Rd4dev commented:
@adhiamboperes, PTAL -- I've addressed the comments, but unsure of handling the emojis.

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

adhiamboperes reviewed (commented):
Thanks @Rd4dev, changes look good to me, and let's discuss the emoji situation.

Rd4dev committed:
Added pull-request trigger to check on every PR creation, removed debug lines and removed the confetti emoji filtering

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

Rd4dev committed:
Trigger only when changes are made to the wiki dir

Rd4dev committed:
Conditional execution of the wiki-deploy to only proceed if the event is push to the develop branch

Rd4dev committed:
Removed the test case for checking TOC present as the log statement to display the TOC not found is removed now

The log statement is removed as it is unnecessary and clutters the check stack trace and also it isn't mandatory to have a TOC on every page so the check is not needed.

Rd4dev commented:
Thanks @adhiamboperes for the review! I’ve added some points on how the current script handles emojis and would like to know if we can proceed with that. Additionally, can we update the existing mismatched headers in this PR to ensure that the check passes completely?

### Mismatching files include
- [Writing tests with good behavioral coverage.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c0dd7ce7fab2822fe7d025bcfa8ce1e836b49665165a4fc69cc0a44f84c69e5a)
- [Oppia Android Code Coverage.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-696bbeaab338ccd0f2816eba3863990fe6fad20ec44ef5bec8475950eea26b76)
- [Interpreting CI Results.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c75dc7d6bc6d57637566435c16252ac968e3115f8fdd42bd3bcd89e0fed845cd)
- [Installing Oppia Android.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c8b1758a251834d70031b628b5753cf9ee20e947703ab68deed7e1b3341c2ed6)

For reference, here is a link to a passing wiki check with all the corrections made: [stack trace](https://github.com/Rd4dev/oppia-android/actions/runs/11162600625/job/31027608237#step:4:58)

Rd4dev committed:
Re-adding the test case to hit coverage but now checking with final result as the toc not found statement is removed

adhiamboperes commented:
> Thanks @adhiamboperes for the review! I’ve added some points on how the current script handles emojis and would like to know if we can proceed with that. Additionally, can we update the existing mismatched headers in this PR to ensure that the check passes completely?
> 
> ### Mismatching files include
> * [Writing tests with good behavioral coverage.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c0dd7ce7fab2822fe7d025bcfa8ce1e836b49665165a4fc69cc0a44f84c69e5a)
> * [Oppia Android Code Coverage.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-696bbeaab338ccd0f2816eba3863990fe6fad20ec44ef5bec8475950eea26b76)
> * [Interpreting CI Results.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c75dc7d6bc6d57637566435c16252ac968e3115f8fdd42bd3bcd89e0fed845cd)
> * [Installing Oppia Android.md](https://github.com/Rd4dev/oppia-android/pull/16/files#diff-c8b1758a251834d70031b628b5753cf9ee20e947703ab68deed7e1b3341c2ed6)
> 
> For reference, here is a link to a passing wiki check with all the corrections made: [stack trace](https://github.com/Rd4dev/oppia-android/actions/runs/11162600625/job/31027608237#step:4:58)

Thanks @Rd4dev! It is totally reasonable to make these fixes here.

Rd4dev committed:
Updating the mis-matched TOC to confirm the TOC check status

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into Issue#5015-AddCIToOppiaAndroidWiki(checkTOC)

Rd4dev committed:
Adding concurrency to wiki actions as it continues to run on subsequent triggers

Rd4dev committed:
Modified the visibilities and removed private fun Kdocs, also updated the end index to span Limit index which previously excluded the last toc element as eOfIdx was calculated as the subList from the start index + 2

Rd4dev committed:
Fix ktlint unexpected spacing before (

Rd4dev committed:
Adding confetti emoji to match with the header link

Rd4dev commented:
Thanks, @adhiamboperes. I’ve updated the existing mismatched TOCs and including the [stack trace](https://github.com/oppia/oppia-android/actions/runs/11259246622/job/31307772003?pr=5382#step:4:58) of the check passing — PTAL

Rd4dev committed:
Replaced checking toc end by blank endline with header tags as they are more accurate than possible duplicate blank lines and also headers presence check would confirm the headers present in the TOC are also available in the wiki page

Rd4dev committed:
Fixed ktlint max line exceeded

adhiamboperes reviewed (approved):
Thanks @Rd4dev! This looks good to me.
